source/common/access_log/access_log_formatter.cc

```diff
@@ -64,14 +66,61 @@ std::string FormatterImpl::format(const Http::HeaderMap& request_headers,
   std::string log_line;
   log_line.reserve(256);
 
-  for (const FormatterPtr& formatter : formatters_) {
-    log_line +=
-        formatter->format(request_headers, response_headers, response_trailers, stream_info);
+  for (const FormatterProviderPtr& provider : providers_) {
+    log_line += provider->format(request_headers, response_headers, response_trailers, stream_info);
   }
 
   return log_line;
 }
 
+JsonFormatterImpl::JsonFormatterImpl(
+    const std::map<const std::string, const std::string>& format_mapping) {
+  for (const auto& pair : format_mapping) {
+    auto providers = AccessLogFormatParser::parse(pair.second);
+
+    // Enforce that each key only has one format specifier in it
```

Comment:
Why do we need this restriction? Why can't leaves have arbitrary format strings?

Reply:
Yeah, that's a good question. I mentioned it in the PR summary too:

> At present, only one specifier per key in the format dictionary is allowed. This is because the whole point of JSON logging is to make logs easily machine-parseable. If you can include multiple formats in the same field, then you'll be right back to parsing those manually

The reason I chose to do this is to make the values sane. I didn't really see the value in doing something like:
`{"protocol_and_duration": "%PROTOCOL% %DURATION%"}`
because during log parsing you'd just have to manually parse that (with a regex or whatever) again.

The other use case would be to put plaintext in there, like:
`{"protocol": "my protocol was %PROTOCOL%"}`, which also doesn't seem particularly useful.

To me, it seemed saner to keep things simple. However this is an arbitrary restriction, and I'm not married to it. If you feel strongly that there should be multiple specifiers per key, let me know and I'll go ahead and lift that.

Reply:
I guess my thinking was that the implementation of general purpose `"%FOO% is %BAR%` is already essentially free, you just reuse the existing access log formatter for leaf entries.

I can see legitimate uses here, e.g. where you are building semi-structured JSON for consumption by tools that don't care about the innermost structure.

It seems easier to just do this, as you have less validation to implement :D

Reply:
I do see that argument, but I feel that having a forcing function to keep logs structured is net win. Additionally, it's easy to add multi-token support in the future, but hard to take it back. 

Similar to disallowing nested dictionaries I propose we stick to single-token leaves for the time being, and then expand that if we see demand for it. As you mentioned it'll be a really simple change if we decide to do it.

Reply:
Ok great, will tackle this
- [ ] relax one formatter requiredment